src/cdk/a11y/focus-monitor/focus-monitor.ts

```diff
@@ -134,6 +133,22 @@ export class FocusMonitor implements OnDestroy {
     this._windowFocusTimeoutId = setTimeout(() => this._windowFocused = false);
   }
 
+  /**
+   * Event listener for `focus` and 'blur' events on the document.
+   * Needs to be an arrow function in order to preserve the context when it gets bound.
+   */
+  private _documentFocusAndBlurListener = (event: FocusEvent) => {
+    const target = event.target as HTMLElement|null;
+    // We need to walk up the ancestor chain in order to support `checkChildren`.
+    for (let el = target; el; el = el.parentElement) {
+      if (event.type === 'focus') {
```

Comment:
Can we extract this if block out of the loop? Instead ahead of time create a `handler` method and then for each element just call `handler(el)`

Reply:
We could, but I don't think it's worth it. It won't improve performance in any way and would just make it less readable.

Reply:
It's not a lot of work, but you are doing extra work by doing the string comparisons repeatedly, and I don't think it hurts the readability to move this logic before the loop

Reply:
We can do either

```ts
    const handler = event.type === 'focus'
      ? (el: HTMLElement) => { this._onFocus(event as FocusEvent, el); }
      : (el: HTMLElement) => { this._onBlur(event as FocusEvent, el); };

    // We need to walk up the ancestor chain in order to support `checkChildren`.
    for (let el = target; el; el = el.parentElement) {
      handler(el);
    }
```

or 

```ts
    const handler = event.type === 'focus' ? this._onFocus : this._onBlur;

    // We need to walk up the ancestor chain in order to support `checkChildren`.
    for (let el = target; el; el = el.parentElement) {
      handler.call(this, event as FocusEvent, el);
    }
```

Both seem to be less readable to me. Performance-wise, comparing the same string just a few times is irrelevant. WDYT?

Reply:
I prefer either option 2 that you gave or:

```ts
const handler = (event.type === 'focus' ? this._onFocus : this._onBlur)
    .bind(this, event as FocusEvent);
```

I don't find them unreadable

Reply:
Done. Also removed an unneeded cast (a leftover from a different approach).